src/splay_tree.jl

```diff
@@ -0,0 +1,231 @@
+        # grand-parent is Null
+        if isa(grand_parent, Nothing)
+            # single rotation
+            if node_x == parent.leftChild
+                # zig rotation
+                right_rotate!(tree, node_x.parent)
+            else 
+                # zag rotation
+                left_rotate!(tree, node_x.parent)
+            end
+            # double rotation
+        elseif node_x == parent.leftChild && parent == grand_parent.leftChild
+            # zig-zig rotation
+            right_rotate!(tree, grand_parent)
+            right_rotate!(tree, parent)
+        elseif node_x == parent.rightChild && parent == grand_parent.rightChild
+            # zag-zag rotation
+            left_rotate!(tree, grand_parent)
+            left_rotate!(tree, parent)
+        elseif node_x == parent.rightChild && parent == grand_parent.leftChild
+            # zig-zag rotation
+            left_rotate!(tree, node_x.parent)
+            right_rotate!(tree, node_x.parent)
+        else
+            # zag-zig rotation
+            right_rotate!(tree, node_x.parent)
+            left_rotate!(tree, node_x.parent)
+        end
+    end
+end
+
+function maximum_node(node::SplayTreeNode_or_null)
+    (node == nothing) && return node
+    while node.rightChild != nothing
+        node = node.rightChild
+    end
+    return node
+end
+
+function _join(tree::SplayTree ,s::SplayTreeNode_or_null, t::SplayTreeNode_or_null)
+    if isa(s, Nothing)
+        return t
+    elseif isa(t, Nothing)
+        return s
+    else
+        x = maximum_node(s)
+        splay!(tree, x)
+        x.rightChild = t
+        t.parent = x
+        return x
+    end
+end
+
+function search_node(tree::SplayTree{K}, d::K) where K
+    node = tree.root
+    prev = nothing
+    while node != nothing && node.data != d
+        prev = node
+        if node.data < d
+            node = node.rightChild
+        else
+            node = node.leftChild
+        end
+    end
+    return (node == nothing) ? prev : node
+end
+
+function haskey(tree::SplayTree{K}, d::K) where K
+    node = tree.root
+    if isa(node, Nothing)
+        return false
+    else
+        node = search_node(tree, d)
+        isa(node, Nothing) && return false
+        is_found = (node.data == d)
+        is_found && splay!(tree, node)
+        return is_found
+    end
+end
+
+
+Base.in(key, tree::SplayTree) = haskey(tree, key)
+
+function Base.delete!(tree::SplayTree{K}, d::K) where K
+    node = tree.root
+    x = search_node(tree, d)
+    (x ==  nothing) && return tree
+    t = nothing
+    s = nothing 
+    
+    splay!(tree, x)
+    
+    if !isa(x.rightChild, Nothing)
+        t = x.rightChild
+        t.parent = nothing
+    end
+
+    s = x
+    s.rightChild = nothing
+
+    if !isa(s.leftChild, Nothing)
+        s.leftChild.parent = nothing
+    end
+
+    tree.root = _join(tree, s.leftChild, t)
+    tree.count -= 1
+    return tree
+end
+
+function Base.insert!(tree::SplayTree{K}, d::K) where K
```

Comment:
We probably should not have 2 argument `insert!`,  just `push!`.
Base only provides a 3 argument `insert` for things with seperate indexes and values.
Base provides it for `Vector` but not for `Set`

Reply:
Possibly we need to deprecate it out of other trees also.
I should have spotted this before